beam-lib/src/http_util.rs

```diff
@@ -1,21 +1,244 @@
+    /// If you don't need any special client configuration use [`BeamClient::new`].
+    pub fn from_client(client: Client, beam_proxy_url: Url) -> Self {
+        Self { client, beam_proxy_url }
+    }
+
+    /// Poll beam tasks using the `filter=todo` option and the given blocking options.
+    /// The generic Parameter T represents the task body type that the requests are expected to have.
+    pub async fn poll_pending_tasks<T: DeserializeOwned>(&self, blocking: &BlockingOptions) -> Result<Vec<TaskRequest<T>>> {
+        let url = self.beam_proxy_url
+            .join(&format!("/v1/tasks?filter=todo&{}", blocking.to_query()))
+            .expect("The proxy url is valid");
+        let response_result = self.client
+            .get(url)
+            .send()
+            .await;
+        let response = match response_result {
+            Ok(res) => res,
+            Err(e) => return if e.is_timeout() {
+                Ok(Vec::with_capacity(0))
+            } else {
+                Err(e.into())
+            },
+        };
+        match response.status() {
+            StatusCode::GATEWAY_TIMEOUT => Ok(Vec::with_capacity(0)),
+            StatusCode::OK | StatusCode::PARTIAL_CONTENT => Ok(response.json().await?),
+            status => Err(BeamError::UnexpectedStatus(status))
+        }
+    }
+
+    /// Poll beam results for a given task id using the given blocking options.
+    /// The generic Parameter T represents the result body type that the requests are expected to have.
+    pub async fn poll_results<T: DeserializeOwned>(&self, task_id: &MsgId, blocking: &BlockingOptions) -> Result<Vec<TaskResult<T>>> {
+        let url = self.beam_proxy_url
+            .join(&format!("/v1/tasks/{task_id}/results?{}", blocking.to_query()))
+            .expect("The proxy url is valid");
+        let response_result = self.client
+            .get(url)
+            .send()
+            .await;
+        let response = match response_result {
+            Ok(res) => res,
+            Err(e) => return if e.is_timeout() {
+                Ok(Vec::with_capacity(0))
+            } else {
+                Err(e.into())
+            },
+        };
+        match response.status() {
+            // TODO: Add more status codes here
+            StatusCode::GATEWAY_TIMEOUT => Ok(Vec::with_capacity(0)),
+            StatusCode::OK | StatusCode::PARTIAL_CONTENT => Ok(response.json().await?),
+            status => Err(BeamError::UnexpectedStatus(status))
+        }
+    }
+
+    /// Post a beam task with a serializeable body.
+    pub async fn post_task<T: Serialize>(&self, task: &TaskRequest<T>) -> Result<()> {
+        let url = self.beam_proxy_url
+            .join("/v1/tasks")
+            .expect("The proxy url is valid");
+        let response = self.client
+            .post(url)
+            .json(task)
+            .send()
+            .await?;
+        match response.status() {
```

Comment:
As above

Reply:
Here it makes a bit more sense but I still disagree. I don't think an if would really read that much nicer.